executor/point_get.go

```diff
@@ -115,22 +119,82 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.RecordBatch) err
 	return e.decodeRowValToChunk(val, req.Chunk)
 }
 
-func (e *PointGetExecutor) encodeIndexKey() ([]byte, error) {
+func (e *PointGetExecutor) encodeIndexKey() (_ []byte, err error) {
+	sc := e.ctx.GetSessionVars().StmtCtx
 	for i := range e.idxVals {
 		colInfo := e.tblInfo.Columns[e.idxInfo.Columns[i].Offset]
-		casted, err := table.CastValue(e.ctx, e.idxVals[i], colInfo)
+		if colInfo.Tp == mysql.TypeString || colInfo.Tp == mysql.TypeVarString || colInfo.Tp == mysql.TypeVarchar {
+			e.idxVals[i], err = handlePadCharToFullLength(sc, &colInfo.FieldType, e.idxVals[i])
+		} else {
+			e.idxVals[i], err = table.CastValue(e.ctx, e.idxVals[i], colInfo)
+		}
 		if err != nil {
 			return nil, err
 		}
-		e.idxVals[i] = casted
 	}
-	encodedIdxVals, err := codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, nil, e.idxVals...)
+
+	encodedIdxVals, err := codec.EncodeKey(sc, nil, e.idxVals...)
 	if err != nil {
 		return nil, err
 	}
 	return tablecodec.EncodeIndexSeekKey(e.tblInfo.ID, e.idxInfo.ID, encodedIdxVals), nil
 }
 
+// handlePadCharToFullLength handles the "PAD_CHAR_TO_FULL_LENGTH" sql mode for
+// CHAR[N] index columns.
+
+// NOTE: kv.ErrNotExist is returned to indicate that this value can not match
+//		 any (key, value) pair in tikv storage. This error should be handle by
+//		 the caller.
+func handlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType, val types.Datum) (types.Datum, error) {
+	targetStr, err := val.ToString()
+	if err != nil {
+		return val, err
+	}
+
+	hasBinaryFlag := mysql.HasBinaryFlag(ft.Flag)
+	isVarchar := ft.Tp == mysql.TypeVarString || ft.Tp == mysql.TypeVarchar
+	isChar := ft.Tp == mysql.TypeString
+	isBinary := ft.Tp == mysql.TypeString && ft.Collate == charset.CollationBin
+	isVarBinary := ft.Tp == mysql.TypeString && ft.Collate == charset.CollationBin
```

Comment:
The conditions of `isVarBinary` it the same as `isBinary`?

Reply:
fixed